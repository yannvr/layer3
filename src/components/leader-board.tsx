import { FC } from 'react';
import useLeaderboardData, { User } from './leader-board.hooks';
import {
  Avatar,
  Container,
  GmStreak,
  Level,
  List,
  ListItem,
  NFTContainerHover,
  RankBadge,
  Stats,
  Title,
  Tooltip,
  UserInfo,
  Username,
  Xp,
} from './leader-board.styles';


interface LeaderBoardProps {
  users?: User[];
  loading?: boolean;
  error?: string | null;
}


const LeaderBoardUser: FC<{ user: User }> = ({ user }) => (
  <ListItem>
    <UserInfo>
      <RankBadge>{user.rank}</RankBadge>
      <Avatar src={`https://ipfs.io/ipfs/${user.avatarCid}`} alt={user.username} />
      <Username href={`https://app.layer3.xyz/{user.username}`} target="_blank">
        {user.username}
      </Username>
    </UserInfo>

    <Stats>
      <GmStreak>{user.gmStreak} GM</GmStreak>
      <Xp>{user.xp} XP</Xp>
      <Level>Lvl {user.level}</Level>
      <div>
        {user.nfts && user.nfts.length > 0 && (
          <NFTContainerHover>
            <img src="/src/assets/nft-icon.png" alt="NFT" width="20" height="20" />
            <Tooltip>
              {user.nfts.map(
                (nft) =>
                  nft.display_image_url && (
                    <div key={nft.identifier}>
                      <a href={nft.opensea_url} target="_blank" rel="noopener noreferrer">
                        <img src={nft.display_image_url} alt={nft.name} width="50" height="50" />
                      </a>
                      <p>{nft.name}</p>
                    </div>
                  ),
              )}
            </Tooltip>
          </NFTContainerHover>
        )}
      </div>
    </Stats>
  </ListItem>
);

export const LeaderBoard: FC<LeaderBoardProps> = ({ users: propUsers, loading: propLoading, error: propError }) => {
  const { users: hookUsers, loading: hookLoading, error: hookError } = useLeaderboardData();
  const users = propUsers ?? hookUsers;
  const loading = propLoading ?? hookLoading;
  const error = propError ?? hookError;

  if (loading) {
    return <Container className="text-center">Loading...</Container>;
  }

  if (error) {
    return <Container className="text-center text-red-500">Error: {error}</Container>;
  }

  return (
    <Container>
      <Title>Leaderboard</Title>
      <List>
        {users.map((user) => (
          <LeaderBoardUser key={user.address} user={user} />
        ))}
      </List>
    </Container>
  );
};
