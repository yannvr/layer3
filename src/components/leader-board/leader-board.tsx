import { FC } from 'react';
import { Link } from 'react-router-dom';
import useLeaderboardData, { User } from './leader-board.hooks';
import {
  Avatar,
  Container,
  GmStreak,
  Level,
  List,
  ListItem,
  RankBadge,
  Stats,
  Title,
  UserInfo,
  Username,
  XP,
} from './leader-board.styles';

interface LeaderBoardProps {
  users?: User[];
  loading?: boolean;
  error?: string | null;
}

const LeaderBoardUser: FC<{ user: User }> = ({ user }) => (
  <ListItem>
    <UserInfo>
      <RankBadge>{user.rank}</RankBadge>
      <Avatar src={`https://ipfs.io/ipfs/${user.avatarCid}`} alt={user.username} />
      <Username as={Link} to={`/user/${user.address}`}>
        {user.username}
      </Username>
    </UserInfo>

    <Stats>
      <GmStreak> ðŸ”¥ {user.gmStreak} </GmStreak>
      <XP> {user.xp} XP </XP>
      <Level>ðŸŽ– {user.level}</Level>
    </Stats>
  </ListItem>
);

export const LeaderBoardRaw: FC<LeaderBoardProps> = ({ users = [], loading = false, error = null }) => {
  if (loading) {
    return <Container className="text-center">Loading...</Container>;
  }

  if (error) {
    return <Container className="text-center text-red-500">Error: {error}</Container>;
  }

  return (
    <Container>
      <Title>Leaderboard</Title>
      <List>
        {users.map((user) => (
          <LeaderBoardUser key={user.address} user={user} />
        ))}
      </List>
    </Container>
  );
};

const LeaderBoard: FC<LeaderBoardProps> = (props) => {
  const { users, loading, error } = useLeaderboardData();
  return <LeaderBoardRaw users={users} loading={loading} error={error} {...props} />;
};

export default LeaderBoard;
