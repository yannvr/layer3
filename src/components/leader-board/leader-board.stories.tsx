// filepath: src/components/leader-board/leader-board.stories.tsx
import type { Meta } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import { LeaderBoardRaw } from './leader-board';
import { LEADER_BOARD_USER_MOCK } from './leader-board.hooks';

export default {
  title: 'Components/LeaderBoard',
  component: LeaderBoardRaw,
  decorators: [
    (Story) => (
      <MemoryRouter initialEntries={['/']}>
        <Story />
      </MemoryRouter>
    ),
  ],
} as Meta;

export const Default = {
  args: {
    users: LEADER_BOARD_USER_MOCK.users,
    loading: false,
    error: null,
  },
};

export const TopThree = {
  args: {
    users: LEADER_BOARD_USER_MOCK.users.slice(0, 3),
    loading: false,
    error: null,
  },
};

export const Loading = {
  args: {
    users: [],
    loading: true,
    error: null,
  },
};

export const Error = {
  args: {
    users: [],
    loading: false,
    error: 'Failed to fetch leaderboard data',
  },
};
