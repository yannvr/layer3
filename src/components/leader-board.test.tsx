import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';
import { LeaderBoard as LeaderBoardRaw } from './leader-board';

const mockUsers: { users: any } = {
  users: [
    {
      rank: 1,
      address: '0xF222f955Ecced246f3181d14fB4629469cEB7681',
      avatarCid: 'QmTUefEyqzfSugwvbCnTjzRdFvp4L5yA6qjEx1yspsr17z',
      username: 'yakugakusei.eth',
      gmStreak: 248,
      xp: 169164,
      level: 41,
    },
    {
      rank: 2,
      address: '0x1234567890abcdef1234567890abcdef12345678',
      avatarCid: 'QmExampleCid2',
      username: 'user2.eth',
      gmStreak: 150,
      xp: 120000,
      level: 30,
    },
  ],
};

describe('LeadBoard Component', () => {
  beforeEach(() => {
    // @ts-ignore
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockUsers),
      })
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('renders loading state initially', () => {
    render(<LeaderBoardRaw />);
    expect(screen.getByText(/loading/i)).toBeInTheDocument();
  });

  test('renders error state when fetch fails', async () => {
    // @ts-ignore
    global.fetch.mockImplementationOnce(() =>
      Promise.resolve({
        ok: false,
      })
    );
    render(<LeaderBoardRaw />);
    await waitFor(() =>
      expect(screen.getByText(/error/i)).toBeInTheDocument()
    );
  });

  test('renders users after successful fetch', async () => {
    render(<LeaderBoardRaw />);
    await waitFor(() =>
      expect(screen.getByText('Leaderboard')).toBeInTheDocument()
    );
    expect(screen.getByText('yakugakusei.eth')).toBeInTheDocument();
    expect(screen.getByText('user2.eth')).toBeInTheDocument();
  });
});
