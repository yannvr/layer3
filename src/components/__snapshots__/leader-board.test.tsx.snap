// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeadBoard Component matches snapshot when populated with NFTs 1`] = `
<DocumentFragment>
  <div
    class="sc-FEMpB tVkqj"
  >
    <h2
      class="sc-gjcoXW evecgK"
    >
      Leaderboard
    </h2>
    <ul
      class="sc-dVBluf cmzpOe"
    >
      <li
        class="sc-fLDLck kxoDan"
      >
        <div
          class="sc-dycHyt gojcEs"
        >
          <span
            class="sc-jbAkgO dfZwmq"
          >
            1
          </span>
          <img
            alt="yakugakusei.eth"
            class="sc-gNZgCX cambCd"
            src="https://ipfs.io/ipfs/QmTUefEyqzfSugwvbCnTjzRdFvp4L5yA6qjEx1yspsr17z"
          />
          <a
            class="sc-kfeOyU PACFW"
            href="https://app.layer3.xyz/{user.username}"
            target="_blank"
          >
            yakugakusei.eth
          </a>
        </div>
        <div
          class="sc-hRDKVd iPvKte"
        >
          <span
            class="sc-jBIHhB jcbsD"
          >
            248 GM
          </span>
          <span
            class="sc-cMqPwU kZRiJs"
          >
            169164 XP
          </span>
          <span
            class="sc-jUkaYT kRKbRX"
          >
            Lvl 41
          </span>
          <div>
            <div
              class="sc-deSjom sc-jZTQcj bpfXyB HJqlG"
            >
              <img
                alt="NFT"
                height="20"
                src="/src/assets/nft-icon.png"
                width="20"
              />
              <div
                class="sc-cSaEAk bhJyLN"
              >
                <div>
                  <a
                    href="https://opensea.io/assets/1"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <img
                      alt="NFT 1"
                      height="50"
                      src="https://example.com/nft1.png"
                      width="50"
                    />
                  </a>
                  <p>
                    NFT 1
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        class="sc-fLDLck kxoDan"
      >
        <div
          class="sc-dycHyt gojcEs"
        >
          <span
            class="sc-jbAkgO dfZwmq"
          >
            2
          </span>
          <img
            alt="user2.eth"
            class="sc-gNZgCX cambCd"
            src="https://ipfs.io/ipfs/QmExampleCid2"
          />
          <a
            class="sc-kfeOyU PACFW"
            href="https://app.layer3.xyz/{user.username}"
            target="_blank"
          >
            user2.eth
          </a>
        </div>
        <div
          class="sc-hRDKVd iPvKte"
        >
          <span
            class="sc-jBIHhB jcbsD"
          >
            150 GM
          </span>
          <span
            class="sc-cMqPwU kZRiJs"
          >
            120000 XP
          </span>
          <span
            class="sc-jUkaYT kRKbRX"
          >
            Lvl 30
          </span>
          <div />
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`LeadBoard Component matches snapshot when populated without NFTs 1`] = `
<DocumentFragment>
  <div
    class="sc-FEMpB tVkqj"
  >
    <h2
      class="sc-gjcoXW evecgK"
    >
      Leaderboard
    </h2>
    <ul
      class="sc-dVBluf cmzpOe"
    >
      <li
        class="sc-fLDLck kxoDan"
      >
        <div
          class="sc-dycHyt gojcEs"
        >
          <span
            class="sc-jbAkgO dfZwmq"
          >
            1
          </span>
          <img
            alt="yakugakusei.eth"
            class="sc-gNZgCX cambCd"
            src="https://ipfs.io/ipfs/QmTUefEyqzfSugwvbCnTjzRdFvp4L5yA6qjEx1yspsr17z"
          />
          <a
            class="sc-kfeOyU PACFW"
            href="https://app.layer3.xyz/{user.username}"
            target="_blank"
          >
            yakugakusei.eth
          </a>
        </div>
        <div
          class="sc-hRDKVd iPvKte"
        >
          <span
            class="sc-jBIHhB jcbsD"
          >
            248 GM
          </span>
          <span
            class="sc-cMqPwU kZRiJs"
          >
            169164 XP
          </span>
          <span
            class="sc-jUkaYT kRKbRX"
          >
            Lvl 41
          </span>
          <div />
        </div>
      </li>
      <li
        class="sc-fLDLck kxoDan"
      >
        <div
          class="sc-dycHyt gojcEs"
        >
          <span
            class="sc-jbAkgO dfZwmq"
          >
            2
          </span>
          <img
            alt="user2.eth"
            class="sc-gNZgCX cambCd"
            src="https://ipfs.io/ipfs/QmExampleCid2"
          />
          <a
            class="sc-kfeOyU PACFW"
            href="https://app.layer3.xyz/{user.username}"
            target="_blank"
          >
            user2.eth
          </a>
        </div>
        <div
          class="sc-hRDKVd iPvKte"
        >
          <span
            class="sc-jBIHhB jcbsD"
          >
            150 GM
          </span>
          <span
            class="sc-cMqPwU kZRiJs"
          >
            120000 XP
          </span>
          <span
            class="sc-jUkaYT kRKbRX"
          >
            Lvl 30
          </span>
          <div />
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
